era_5g_interface/channels.py

```diff
@@ -0,0 +1,294 @@
+        """
+
+        if type(self._sio) is socketio.Client:
+            return str(self._sio.sid)
+        else:
+            return str(self._sio.manager.eio_sid_from_sid(sid, namespace))
+
+    def data_error_callback(self, data: Dict[str, Any], sid: Optional[str] = None) -> None:
+        """Allows to receive general error data on DATA_NAMESPACE.
+
+        Args:
+            data (Dict[str, Any]): JSON data.
+            sid (str, optional): Namespace sid - only on the server side.
+        """
+
+        logger.error(f"Data error, eio_sid {self.get_client_eio_sid(sid, DATA_NAMESPACE)}, sid {sid}, data {data}")
+
+    def image_decode(self, data: Dict[str, Any], event: str, sid: Optional[str] = None) -> Optional[dict]:
+        """Allows to receive jpeg or h264 encoded image using the websocket transport.
+
+        Args:
+            data (Dict[str, Any]): Received dictionary with frame data.
+            event (str): Event name.
+            sid (str, optional): Namespace sid - only on the server side.
+        """
+
+        if "timestamp" in data:
+            timestamp = data["timestamp"]
+        else:
+            logger.info("Timestamp not set, setting default value")
+            timestamp = 0
+
+        if "frame" not in data:
+            logger.error("Data does not contain frame.")
+            self.send_data(
+                {"timestamp": timestamp, "error": "Data does not contain frame."},
+                self._callbacks_info[event].error_event,
+                sid=sid,
+            )
+            return None
+
+        eio_sid = self.get_client_eio_sid(sid, DATA_NAMESPACE)
+        if self._callbacks_info[event].type is ChannelType.H264 and eio_sid not in self._decoders:
+            self._decoders[eio_sid] = H264Decoder(data["width"], data["height"])
+
+        if eio_sid in self._decoders:
+            last_timestamp = self._decoders[eio_sid].last_timestamp
+            if timestamp - last_timestamp < 0:
+                logger.error(
+                    f"Received frame with older timestamp: {timestamp}, "
+                    f"last_timestamp: {last_timestamp}, diff: {timestamp - last_timestamp}"
+                )
+                self.send_data(
+                    {
+                        "timestamp": timestamp,
+                        "error": f"Received frame with older timestamp: {timestamp}, "
+                        f"last_timestamp: {last_timestamp}, diff: {timestamp - last_timestamp}",
+                    },
+                    self._callbacks_info[event].error_event,
+                    sid=sid,
+                )
+                return None
+            self._decoders[eio_sid].last_timestamp = timestamp
+
+        try:
+            if eio_sid in self._decoders:
+                frame_decoded = self._decoders[eio_sid].decode_packet_data(data["frame"])
+            else:
+                frame_decoded = cv2.imdecode(np.frombuffer(data["frame"], dtype=np.uint8), cv2.IMREAD_COLOR)
```

Comment:
...and some CV2 exception (or whatever it raises) here

Reply:
Again, I don't understand.